File: src-tauri/src/commands/submodule.rs
use tauri::State;

use crate::commands::with_repo;
use crate::git::types::SubmoduleInfo;
use crate::state::AppState;

#[tauri::command]
pub fn list_submodules(
    tab_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<SubmoduleInfo>, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.list_submodules().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn add_submodule(
    tab_id: String,
    url: String,
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend
            .add_submodule(&url, &path)
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn update_submodule(
    tab_id: String,
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.update_submodule(&path).map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn update_all_submodules(tab_id: String, state: State<'_, AppState>) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.update_all_submodules().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn remove_submodule(
    tab_id: String,
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.remove_submodule(&path).map_err(|e| e.to_string())
    })
}

File: src/services/submodule.ts
import { invoke } from "@tauri-apps/api/core";

export type SubmoduleStatus =
  | "up_to_date"
  | "modified"
  | "uninitialized"
  | "conflict";

export interface SubmoduleInfo {
  path: string;
  url: string;
  branch: string | null;
  head_oid: string | null;
  head_short_oid: string | null;
  status: SubmoduleStatus;
}

export function listSubmodules(tabId: string): Promise<SubmoduleInfo[]> {
  return invoke<SubmoduleInfo[]>("list_submodules", { tabId });
}

export function addSubmodule(
  tabId: string,
  url: string,
  path: string,
): Promise<void> {
  return invoke<void>("add_submodule", { tabId, url, path });
}

export function updateSubmodule(tabId: string, path: string): Promise<void> {
  return invoke<void>("update_submodule", { tabId, path });
}

export function updateAllSubmodules(tabId: string): Promise<void> {
  return invoke<void>("update_all_submodules", { tabId });
}

export function removeSubmodule(tabId: string, path: string): Promise<void> {
  return invoke<void>("remove_submodule", { tabId, path });
}
